pkg/bgpv1/manager/reconcilerv2/service.go

```diff
@@ -0,0 +1,535 @@
+	if err != nil {
+		return err
+	}
+
+	ls, err := r.populateLocalServices(p.CiliumNode.Name)
+	if err != nil {
+		return fmt.Errorf("failed to populate local services: %w", err)
+	}
+
+	var desiredSvcPaths ServiceAFPathsMap
+	if r.modifiedServiceAdvertisements(p, desiredPeerAdverts) {
+		// BGP configuration for service advertisement changed, we should reconcile all services.
+		desiredSvcPaths, err = r.fullReconciliation(p, ls)
+		if err != nil {
+			return err
+		}
+	} else {
+		// BGP configuration is unchanged, only reconcile modified services.
+		desiredSvcPaths, err = r.svcDiffReconciliation(p, ls)
+		if err != nil {
+			return err
+		}
+	}
+
+	var allErr error
+	for svc, desiredAFPaths := range desiredSvcPaths {
+		metadata := r.getMetadata(p.BGPInstance)
+
+		// check if service exists
+		currentAFPaths, exists := metadata.ServicePaths[svc]
+		if !exists && len(desiredAFPaths) == 0 {
+			// service does not exist in our local state, and there is nothing to advertise
+			continue
+		}
+
+		// reconcile service paths
+		updatedAFPaths, err := ReconcileAFPaths(&ReconcileAFPathsParams{
+			Logger:       r.logger.WithField(types.InstanceLogField, p.DesiredConfig.Name),
+			Ctx:          ctx,
+			Instance:     p.BGPInstance,
+			DesiredPaths: desiredAFPaths,
+			CurrentPaths: currentAFPaths,
+		})
+
+		if err == nil && len(desiredAFPaths) == 0 {
+			// no error is reported and desiredAFPaths is empty, we should delete the service
+			delete(metadata.ServicePaths, svc)
+		} else {
+			// update service paths with returned updatedAFPaths even if there was an error.
+			metadata.ServicePaths[svc] = updatedAFPaths
+		}
+
+		r.setMetadata(p.BGPInstance, metadata)
+		allErr = errors.Join(allErr, err)
+	}
+
+	return allErr
+}
+
+// modifiedServiceAdvertisements compares local advertisement state with desiredPeerAdverts, if they differ, it updates the local state and returns true
+// for full reconciliation.
+func (r *ServiceReconciler) modifiedServiceAdvertisements(p ReconcileParams, desiredPeerAdverts PeerAdvertisements) bool {
+	// current metadata
+	serviceMetadata := r.getMetadata(p.BGPInstance)
+
+	// check if BGP advertisement configuration modified
+	modified := !PeerAdvertisementsEqual(serviceMetadata.ServiceAdvertisements, desiredPeerAdverts)
+
+	// update local state, if modified
+	if modified {
+		r.setMetadata(p.BGPInstance, ServiceReconcilerMetadata{
+			ServicePaths:          serviceMetadata.ServicePaths,
+			ServiceAdvertisements: desiredPeerAdverts,
+		})
+	}
+
+	return modified
+}
+
+// Populate locally available services used for externalTrafficPolicy=local handling
+func (r *ServiceReconciler) populateLocalServices(localNodeName string) (sets.Set[resource.Key], error) {
+	ls := sets.New[resource.Key]()
+
+	epList, err := r.epDiffStore.List()
+	if err != nil {
+		return nil, fmt.Errorf("failed to list upsertEPs from diffstore: %w", err)
```

Comment:
Looks like an IDE replace action renamed `EPs` to `upsertEPs` on multiple places where it was not intended, do a search for it in this file.